### DISCUSSION TITLE
Use nitro server within an external server

### DISCUSSION THREAD

**Discussion by Maxttier** [nitrojs/nitro#2959] [Q&A]
No something to import wich will do all that form me, something like createNitroApp(params)?

```
import { listen } from 'listhen'
import {
  createApp,
  createRouter,
  H3EventContext,
  splitCookiesString,
  isEvent,
  getResponseStatus,
  getResponseStatusText,
  setResponseHeader,
  setResponseStatus,
  send,
  toNodeListener,
  EventHandlerRequest,
  H3Error,
  H3Event
} from 'h3'
import { createHooks } from 'hookable'
import { Headers } from 'ofetch'
import {
  createCall,
  createFetch as createLocalFetch,
  FetchOptions
} from 'unenv/runtime/fetch/index'


  function errorHandler(error: H3Error<unknown>, event: H3Event<EventHandlerRequest>) {
    setResponseHeader(event, 'Content-Type', 'text/html; charset=UTF-8');
    setResponseStatus(event, 503, 'Server Unavailable');

    let body: string;
    let title: string;

    if (error) {
      title = `${getResponseStatus(event)} ${getResponseStatusText(event)}`;
      body = `<code><pre>${error.stack}</pre></code>`;
    } else {
      title = 'Reloading server...';
      body = '<progress></progress><script>document.querySelector("progress").indeterminate=true<\/script>';
    }

    return send(
      event,
      `<!DOCTYPE html>
        <html lang='en'>
        <head>
          <meta charset='utf-8'>
          <meta name='viewport' content='width=device-width, initial-scale=1'>
          ${error ? '' : '<meta http-equiv="refresh" content="2">'}
          <title>${title}</title>
          <link rel='stylesheet' href='https://cdn.jsdelivr.net/npm/@picocss/pico/css/pico.min.css'>
        </head>
        <body>
          <main class='container'>
            <article>
              <header>
                <h2>${title}</h2>
              </header>
              ${body}
              <footer>
                Check console logs for more information.
              </footer>
            </article>
        </main>
        </body>
      </html>
      `
    );
  }

  function joinHeaders(value: number | string | string[]) {
    return Array.isArray(value) ? value.join(', ') : String(value);
  }

  function normalizeCookieHeader(header: number | string | string[] = '') {
    return splitCookiesString(joinHeaders(header));
  }

  function normalizeCookieHeaders(headers: Headers) {
    const outgoingHeaders = new Headers();
    for (const [name, header] of headers) {
      if (name === 'set-cookie') {
        for (const cookie of normalizeCookieHeader(header)) {
          outgoingHeaders.append('set-cookie', cookie);
        }
      } else {
        outgoingHeaders.set(name, joinHeaders(header));
      }
    }
    return outgoingHeaders;
  }

  function normalizeFetchResponse(response: Response) {
    if (!response.headers.has('set-cookie')) {
      return response;
    }
    return new Response(response.body, {
      status: response.status,
      statusText: response.statusText,
      headers: normalizeCookieHeaders(response.headers),
    });
  }

  // const nitroApp = useNitroApp()
  const h3App = createApp({
    debug: true,
    onError: (error, event) => {
      captureError(error, { event, tags: ['request'] });
      return errorHandler(error, event);
    },
    onRequest: async (event) => {
      await nitroApp.hooks.callHook('request', event).catch((error) => {
        captureError(error, { event, tags: ['request'] });
      });
    },
    onBeforeResponse: async (event, response) => {
      await nitroApp.hooks.callHook('beforeResponse', event, response).catch((error) => {
        captureError(error, { event, tags: ['request', 'response'] });
      });
    },
    onAfterResponse: async (event, response) => {
      await nitroApp.hooks.callHook('afterResponse', event, response).catch((error) => {
        captureError(error, { event, tags: ['request', 'response'] });
      });
    }
  });
  
  //@ts-ignore
  const localCall = createCall(toNodeListener(h3App));  // something is bad 
  const _localFetch = createLocalFetch(localCall, globalThis.fetch);
  const localFetch = (input: string | Request, init: FetchOptions) => _localFetch(input, init).then(
    (response) => normalizeFetchResponse(response)
  );

  const hooks = createHooks()
  const router = createRouter({
    preemptive: true
  });
  h3App.use(router)

  const captureError = (error: unknown, context: H3EventContext = {}) => {
    const promise = hooks.callHookParallel('error', error, context).catch((error_) => {
      console.error('Error while capturing another error', error_);
    });
    if (context.event && isEvent(context.event)) {
      const errors = context.event.context.nitro?.errors;
      if (errors) {
        errors.push({ error, context });
      }
      if (context.event.waitUntil) {
        context.event.waitUntil(promise);
      }
    }
  };

  const nitroApp = {
    h3App,
    hooks,
    router,
    localCall,
    localFetch,
    captureError,
  }


  await listen(toNodeListener(h3App))
```

**Comment by Maxttier** [comment]
useNitroApp is not what I need because i don't run `nitro dev` so all magic like `#imports` not work.

Can't import things from nitropack if in tsconfig is set to compile to commonjs

**Comment by Maxttier** [comment]
@pi0  that I meanin https://github.com/nitrojs/nitro/issues/2958

**Comment by pi0** [comment]
@Makkalay can you please either in plain English or your native language if more comfortable with, explain what you need to achieve? if not using nitro server, you can use h3 directly, if using nitro, you need to use `nitro dev` / `nitro build` commands.

**Comment by Maxttier** [comment]
https://github.com/nitrojs/nitro/issues/2957#issuecomment-2563660126

And I sent a request in Discord to you, few years it stay in pending status :)

**Reply by pi0** [reply]
few years probably expired.. please send (a message) again.

**Comment by Maxttier** [comment]
@pi0 I can't cancel request in discord i think

I'm trying to replace a server engine used by some framework with the nitro server engine, and everything is fine. 
I start nitro and pass it to that framework. Everything is fine. 
Eg: in nitro plugin i call new Framework(nitroApp);

Now I need/want to do the opposite, start that framework (not nuxt) and pass nitro to it. 
To do that, I need to create a nitroApp (not h3, because I want to be as close to the first approach as possible). 
Eg: in that framework i call new Framework(nitroApp); when it is initialized.

**Reply by Maxttier** [reply]
With code that I write on question it also work, at least I get responses on api call but is much code and not all is ok i think minimum I get types errors.
Also import from nitropack not work, when build app say no errors and after say what nitropack not export things, in vs code all is ok. Compile to commonjs

**Reply by pi0** [reply]
thanks for more explanations. 

Unfortunately what you are trying to do is not possible at least for development because nitro uses an isolated dev server and expects to be the "main" server. opposite is possible to run any other server within it.

For development, you may run nitro/nuxt in parallel and use a proxy like [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware) to send requests to that server.

For production, however, it is quite easy to do. You can build your nitro/nuxt app with `npm run build --preset node-listener`. You can use `handler(req, res)` export from `.output/server/index.mjs` which is a standard node middleware you can import it in any other custom framework.

If your external server is CommonJS, that's fine also, you still can use `import('.output/server/index.mjs')`

**Reply by Maxttier** [reply]
That framework has its own npm commands an will be nice to use it, and it have its own dev server at least it resbuild on code changes

**Reply by pi0** [reply]
nitro dev server also watches for file change. like I mentioned above, if you really really want to have two dev servers, solution is proxy.

**Reply by Maxttier** [reply]
don't want 2, want somethink like I write in question

**Reply by pi0** [reply]
it is simply NOT possible.

**Reply by Maxttier** [reply]
Any way, how I did it works, so far it is enough to test the behavior of one and the other variant. Thank

**Reply by pi0** [reply]
no problems and best luck.
